test/test_shared_library.cpp

```diff
@@ -0,0 +1,88 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+
+#include <string>
+
+#include "./allocator_testing_utils.h"
+
+#include "rcutils/allocator.h"
+#include "rcutils/error_handling.h"
+#include "rcutils/shared_library.h"
+
+#include "rcutils/get_env.h"
+
+class TestSharedLibrary : public ::testing::Test
+{
+protected:
+  void SetUp() final
+  {
+    // Reset rcutil error global state in case a previously
+    // running test has failed.
+    rcutils_reset_error();
+    lib = rcutils_get_zero_initialized_shared_library();
+  }
+  rcutils_shared_library_t lib;
+};
+
+TEST_F(TestSharedLibrary, basic_load) {
+  rcutils_ret_t ret;
+
+  // checking rcutils_get_zero_initialized_shared_library
+  ASSERT_STRNE(lib.library_path, "");
+  EXPECT_TRUE(lib.lib_pointer == NULL);
+
+  // library path
+  const std::string library_path = std::string("libdummy_shared_library.so");
+
+  // getting shared library
+  ret = rcutils_load_shared_library(&lib, library_path.c_str(), rcutils_get_default_allocator());
+  ASSERT_EQ(RCUTILS_RET_OK, ret);
+
+  // unload shared_library
+  ret = rcutils_unload_shared_library(&lib);
+  ASSERT_EQ(RCUTILS_RET_OK, ret);
+
+  // checking if we have unloaded and freed memory
+  ASSERT_STRNE(lib.library_path, "");
+  EXPECT_TRUE(lib.lib_pointer == NULL);
+}
+
+TEST_F(TestSharedLibrary, basic_symbol) {
+  void * symbol;
+  bool ret;
+
+  symbol = rcutils_get_symbol(nullptr, "symbol");
+  EXPECT_TRUE(symbol == NULL);
+
+  ret = rcutils_has_symbol(nullptr, "symbol");
+  EXPECT_FALSE(ret);
+
+  const std::string library_path = std::string("libdummy_shared_library.so");
+
+  // getting shared library
+  ret = rcutils_load_shared_library(&lib, library_path.c_str(), rcutils_get_default_allocator());
+  ASSERT_EQ(RCUTILS_RET_OK, ret);
+
+  symbol = rcutils_get_symbol(&lib, "symbol");
+  EXPECT_TRUE(symbol == NULL);
+
+  symbol = rcutils_get_symbol(&lib, "print_name");
+  EXPECT_TRUE(symbol != NULL);
+
+  // unload shared_library
+  ret = rcutils_unload_shared_library(&lib);
+  ASSERT_EQ(RCUTILS_RET_OK, ret);
+}
```

Comment:
The tests should also exercise non-standard sequences of function calls, e.g. load twice, get_symbol before loading, unload twice, etc.

Reply:
I added more tests https://github.com/ros2/rcutils/pull/215/commits/791214afc1a389a57bde249956c9efdd52c77676

Reply:
E.g. the case of calling `rcutils_load_shared_library()` twice isn't covered yet. And I assume the current implementation would leak memory in that case.

Reply:
Avoided the leak https://github.com/ros2/rcutils/pull/215/commits/8dc66df8868220f0d3052d1b3e560ac20275b1dc and added the test